src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,749 @@
+    }
+  }
+
+
+  def minimal(tt: TableType): TableType = {
+    val keySet = tt.key.iterator.flatten.toSet
+    tt.copy(
+      rowType = tt.rowType.filter(keySet)._1,
+      globalType = TStruct(tt.globalType.required)
+    )
+  }
+
+  def minimal(mt: MatrixType): MatrixType = {
+    val rowKeySet = mt.rowKey.toSet
+    val colKeySet = mt.colKey.toSet
+    mt.copy(
+      rvRowType = mt.rvRowType.filter(rowKeySet ++ Array(MatrixType.entriesIdentifier))._1,
+      colType = mt.colType.filter(colKeySet)._1,
+      globalType = TStruct(mt.globalType.required)
+    )
+  }
+
+  def minimal[T <: Type](base: T): T = {
+    val result = base match {
+      case ts: TStruct => TStruct(ts.required)
+      case ta: TArray => TArray(minimal(ta.elementType), ta.required)
+      case tagg: TAggregable => tagg.copy(elementType = minimal(tagg.elementType))
+      case t => t
+    }
+    result.asInstanceOf[T]
+  }
+
+  def unionBaseType[T <: BaseType](base: T, children: T*): T = {
+    val r = base match {
+      case tt: TableType =>
+        val ttChildren = children.map(_.asInstanceOf[TableType])
+        tt.copy(
+          rowType = unionType(tt.rowType, ttChildren.map(_.rowType): _*),
+          globalType = unionType(tt.globalType, ttChildren.map(_.globalType): _*)
+        )
+      case mt: MatrixType =>
+        val mtChildren = children.map(_.asInstanceOf[MatrixType])
+        mt.copy(
+          globalType = unionType(mt.globalType, mtChildren.map(_.globalType): _*),
+          rvRowType = unionType(mt.rvRowType, mtChildren.map(_.rvRowType): _*),
+          colType = unionType(mt.colType, mtChildren.map(_.colType): _*)
+        )
+    }
+    r.asInstanceOf[T]
+  }
+
+  def unionType[T <: Type](base: T, children: T*): T = {
+    children.foreach(t => assert(isSupertype(t, base), s"incompatibility:\n  ${ t.parsableString() }\n  ${ base.parsableString }"))
+
+    if (children.isEmpty)
+      return minimal(base)
+
+    val r = base match {
+      case ts: TStruct =>
+        val subStructs = children.map(_.asInstanceOf[TStruct])
+        val subFields = ts.fields.map { f =>
+          f -> subStructs.flatMap(s => s.fieldOption(f.name))
+        }
+          .filter(_._2.nonEmpty)
+          .map { case (f, ss) => f.name -> unionType(f.typ, ss.map(_.typ): _*) }
+        TStruct(ts.required, subFields: _*)
+      case ta: TArray =>
+        TArray(unionType(ta.elementType, children.map(_.asInstanceOf[TArray].elementType): _*), ta.required)
+      case tagg: TAggregable => tagg.copy(elementType =
+        unionType(tagg.elementType, children.map(_.asInstanceOf[TAggregable].elementType): _*))
+      case _ =>
+        assert(children.forall(_ == base))
+        base
+    }
+    r.asInstanceOf[T]
+  }
+
+  def unify(base: Env[Type], envs: Env[Type]*): Env[Type] = {
+    val bindings = base.m.toArray.map { case (name, baseType) =>
+      name -> unionType[Type](baseType, envs.flatMap(_.lookupOption(name)): _*)
```

Comment:
This might be clear as I read further, but why would the names not all have the same bindings?  Is env special?  Or is the intersection of names useful here?